pkg/pool-manager/pool.go

```diff
@@ -186,31 +194,78 @@ func getNextMac(currentMac net.HardwareAddr) net.HardwareAddr {
-func (p *PoolManager) isInstanceOptedIn(namespaceName, mutatingWebhookConfigName, webhookName string) (bool, error) {
+// Checks if the namespace of an instance needs to managed by kubemacpool
+func (p *PoolManager) compareNamespaceSelectorToOptMode(namespaceName, mutatingWebhookConfigName, webhookName string, vmOptMode OptMode) (bool, error) {
+	var isNamespaceManaged bool
+	if vmOptMode == OptInMode {
+		// in OptInMode namespaces are not managed by default
+		isNamespaceManaged = false
+	} else if vmOptMode == OptOutMode {
+		// in OptOutMode namespaces are managed by default
+		isNamespaceManaged = true
+	} else {
+		return false, fmt.Errorf("opt-mode is not defined: %s", vmOptMode)
+	}
+
 	ns, err := p.kubeClient.CoreV1().Namespaces().Get(context.TODO(), namespaceName, metav1.GetOptions{})
 	if err != nil {
-		return false, errors.Wrapf(err, "Failed to get Namespace %s", namespaceName)
+		return isNamespaceManaged, errors.Wrapf(err, "Failed to get Namespace %s", namespaceName)
 	}
 	namespaceLabelMap := ns.GetLabels()
-	log.V(3).Info("namespaceName Labels", "namespaceName", namespaceName, "Labels", namespaceLabelMap)
+	log.V(1).Info("namespaceName Labels", "vm instance namespaceName", namespaceName, "Labels", namespaceLabelMap)
 
 	mutatingWebhookConfiguration, err := p.kubeClient.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Get(context.TODO(), mutatingWebhookConfigName, metav1.GetOptions{})
 	if err != nil {
-		return false, errors.Wrapf(err, "Failed to get mutatingWebhookConfig %s", mutatingWebhookConfigName)
+		return isNamespaceManaged, errors.Wrapf(err, "Failed to get mutatingWebhookConfig %s", mutatingWebhookConfigName)
 	}
 
 	for _, webhook := range mutatingWebhookConfiguration.Webhooks {
 		if webhook.Name == webhookName {
 			if namespaceLabelSet := labels.Set(namespaceLabelMap); namespaceLabelSet != nil {
-				if webhookNamespaceLabelSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector); err == nil && webhookNamespaceLabelSelector.Matches(namespaceLabelSet) {
-					return true, nil
+				webhookNamespaceLabelSelector, err := metav1.LabelSelectorAsSelector(webhook.NamespaceSelector)
+				if err != nil {
+					return isNamespaceManaged, errors.Wrapf(err, "Failed to set webhook Namespace Label Selector for webhook %s", webhook.Name)
 				}
+
+				log.V(1).Info("webhook NamespaceLabelSelectors", "webhookName", webhookName, "NamespaceLabelSelectors", webhookNamespaceLabelSelector)
+				if vmOptMode == OptInMode && webhookNamespaceLabelSelector.Matches(namespaceLabelSet) {
+					// if we are in opt-in mode then we check that the namespace has the including label
+					isNamespaceManaged = true
+				} else if vmOptMode == OptOutMode && !webhookNamespaceLabelSelector.Matches(namespaceLabelSet) {
+					// if we are in opt-out mode then we check that the namespace does not have the excluding label
+					isNamespaceManaged = false
+				}
```

Comment:
This looks like a different check for the namespace-manahed. If before it was checked based on the  opt-mode defaults, not it is checked against the label selector.
So, it will be nice to extract it to its own function, naming it per what it does so it will be easy to read without entering the low details.

In addition, the boolean is updated here on every webhook, each one overwriting the previous. Was this the intention (it is not "any" nor "all" operation)?
Maybe you depend on entering this area only once per the webhook name?

Reply:
> This looks like a different check for the namespace-manahed. If before it was checked based on the opt-mode defaults, not it is checked against the label selector.
> So, it will be nice to extract it to its own function, naming it per what it does so it will be easy to read without entering the low details.

I agree

> 
> In addition, the boolean is updated here on every webhook, each one overwriting the previous. Was this the intention (it is not "any" nor "all" operation)?
> Maybe you depend on entering this area only once per the webhook name?

IIUC the boolean will only be updated if it matches the name, which is unique, so it should only happen once (if any).